plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -969,6 +993,83 @@ private void expireSnapshots(Table table, long expireTimestamp, ConnectorSession
+    }
+
+    private void deleteOrphanFiles(Table table, ConnectorSession session, SchemaTableName schemaTableName, long expireTimestamp)
+    {
+        Set<String> validDataFilePaths = stream(table.snapshots())
+                .map(Snapshot::snapshotId)
+                .flatMap(snapshotId -> stream(table.newScan().useSnapshot(snapshotId).planFiles()))
+                .map(fileScanTask -> URI.create(fileScanTask.file().path().toString()).getPath())
+                .collect(toImmutableSet());
+        scanAndDeleteInvalidFiles(table, session, schemaTableName, expireTimestamp, validDataFilePaths, "/data");
+    }
+
+    private void deleteOrphanMetadataFiles(Table table, ConnectorSession session, SchemaTableName schemaTableName, long expireTimestamp)
+    {
+        ImmutableSet<String> manifests = stream(table.snapshots())
+                .flatMap(snapshot -> snapshot.allManifests().stream())
+                .map(ManifestFile::path)
+                .collect(toImmutableSet());
+        List<String> manifestLists = ReachableFileUtil.manifestListLocations(table);
+        List<String> otherMetadataFiles = concat(
+                metadataFileLocations(table, false).stream(),
+                Stream.of(versionHintLocation(table)))
+                .collect(toImmutableList());
+        Set<String> validMetadataFiles = concat(manifests.stream(), manifestLists.stream(), otherMetadataFiles.stream())
+                .map(path -> URI.create(path).getPath())
+                .collect(toImmutableSet());
+        scanAndDeleteInvalidFiles(table, session, schemaTableName, expireTimestamp, validMetadataFiles, "/metadata");
+    }
+
+    private void scanAndDeleteInvalidFiles(Table table, ConnectorSession session, SchemaTableName schemaTableName, long expireTimestamp, Set<String> validFiles, String subfolder)
+    {
+        try {
+            FileSystem fileSystem = hdfsEnvironment.getFileSystem(new HdfsEnvironment.HdfsContext(session), new Path(table.location()));
+            RemoteIterator<LocatedFileStatus> allFiles = fileSystem.listFiles(new Path(table.location() + subfolder), true);
+            while (allFiles.hasNext()) {
+                LocatedFileStatus file = allFiles.next();
+                if (file.isFile()) {
+                    String normalizedPath = file.getPath().toUri().getPath();
+                    if (file.getModificationTime() < expireTimestamp && !validFiles.contains(normalizedPath)) {
+                        fileSystem.delete(file.getPath(), false);
+                        log.debug("%s file deleted while deleting orphan files %s", file.getPath().toString(), schemaTableName.getTableName());
+                    }
+                    else {
+                        log.debug("%s file retained while deleting orphan files %s", file.getPath().toString(), schemaTableName.getTableName());
+                    }
+                }
+                else if (file.isDirectory()) {
```

Comment:
Should we delete directories? What's the use-case?
For partitioned data, someone may be about to write to a currently obsolete partition directory, and thus this may make the write fail